src/main/kotlin/org/dif/operations/Unpack.kt

```diff
@@ -0,0 +1,110 @@
+package org.dif.operations
+
+import org.dif.common.AnonCryptAlg
+import org.dif.common.AuthCryptAlg
+import org.dif.crypto.ParseResult
+import org.dif.crypto.anonDecrypt
+import org.dif.crypto.authDecrypt
+import org.dif.crypto.getCryptoAlg
+import org.dif.crypto.key.RecipientKeySelector
+import org.dif.crypto.parse
+import org.dif.crypto.verify
+import org.dif.exceptions.MalformedMessageException
+import org.dif.message.Message
+import org.dif.model.Metadata
+import org.dif.model.UnpackParams
+import org.dif.model.UnpackResult
+import org.dif.utils.component1
+import org.dif.utils.component2
+
+fun unpack(params: UnpackParams, keySelector: RecipientKeySelector): UnpackResult {
+    val metadataBuilder = Metadata.Builder()
+
+    val msg = when (val parseResult = parse(params.packedMessage)) {
+        is ParseResult.JWS -> parseResult.unpack(keySelector, metadataBuilder)
+        is ParseResult.JWE -> parseResult.unpack(keySelector, metadataBuilder)
+        is ParseResult.JWM -> parseResult.message
+    }
+
+    return UnpackResult(msg, metadataBuilder.build())
+}
+
+private fun ParseResult.JWS.unpack(keySelector: RecipientKeySelector, metadataBuilder: Metadata.Builder): Message {
+    val kid = message.unprotectedHeader?.keyID
+        ?: throw MalformedMessageException("JWS Unprotected Per-Signature header must be present")
+
+    val key = keySelector.verifyKey(kid)
```

Comment:
`verifyKey` sounds ambiguous (looks like we do verification of a key). Maybe call it `getVerificationKey`?  (same for `signKey` in `SenderKeySelector`, I would call it `getSigningKey`)